File: backend/app/schemas/video.py
from typing import Optional, Dict, Any, List
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict

class SmartAutoplaySettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    enabled: bool = False
    mode: Optional[str] = "smart"  # 'smart' (chamada para desmutar) ou 'direct' (autoplay direto com som)
    text: str = "Seu vídeo já começou!"
    subtext: str = "Clique no botão abaixo para ativar o som"
    button_color: str = "#ef4444"
    button_text: str = "CLIQUE PARA OUVIR"
    restart_on_unmute: bool = False
    size: Optional[str] = "medium"

class FloatingPlayerSettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    enabled: bool = False
    position: str = "bottom-right"
    width: int = 320
    closeable: bool = True

class PitchDelaySettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    enabled: bool = False
    time: int = 60
    target_css_selector: str = ".delay-pitch"
    auto_scroll: bool = True
    scroll_offset: int = 50
    persistence: bool = True

class PixelEventConfig(BaseModel):
    model_config = ConfigDict(extra="allow")
    trigger: str
    event_name: str
    enabled: bool = True

class TrackingPixelsSettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    enabled: bool = False
    facebook_pixel_id: Optional[str] = None
    google_analytics_id: Optional[str] = None
    tiktok_pixel_id: Optional[str] = None
    events: Optional[List[PixelEventConfig]] = None

class DomainProtectionSettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    enabled: bool = False
    allowed_domains: List[str] = Field(default_factory=list)
    anti_download: bool = True

class PlayerControlsConfig(BaseModel):
    model_config = ConfigDict(extra="allow")
    rewind_10s: bool = True
    forward_10s: bool = True
    volume: bool = True
    fullscreen: bool = True
    speed_control: bool = True
    progress_bar: bool = True
    video_time: bool = True

class ChapterItem(BaseModel):
    model_config = ConfigDict(extra="allow")
    id: str
    time: str = "00:00"
    seconds: float = 0.0
    title: str = ""

class ChaptersSettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    enabled: bool = False
    items: List[ChapterItem] = Field(default_factory=list)

class PlayerSettings(BaseModel):
    model_config = ConfigDict(extra="allow")
    primary_color: str = "#6366f1"
    autoplay: bool = False
    show_controls: bool = True
    border_radius: Optional[int] = 0
    aspect_ratio: Optional[str] = "16:9"
    cta_enabled: bool = False
    cta_time: int = 0
    cta_text: str = "Comprar Agora"
    cta_link: str = "https://example.com"
    turbo_enabled: Optional[bool] = False
    playback_rate: Optional[float] = 1.0
    smart_autoplay: Optional[SmartAutoplaySettings] = None
    floating_player: Optional[FloatingPlayerSettings] = None
    pitch_delay: Optional[PitchDelaySettings] = None
    tracking_pixels: Optional[TrackingPixelsSettings] = None
    domain_protection: Optional[DomainProtectionSettings] = None
    controls_config: Optional[PlayerControlsConfig] = None
    chapters: Optional[ChaptersSettings] = None
    transparent_background: Optional[bool] = False
    remove_black_bars: Optional[bool] = True
    fit_mode: Optional[str] = "cover"



class VideoCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    video_url: str = Field(..., min_length=1, max_length=1024)
    thumbnail_url: Optional[str] = Field(None, max_length=1024)
    duration: Optional[float] = 0.0
    player_settings: Optional[PlayerSettings] = None

class VideoUpdate(BaseModel):
    title: Optional[str] = None
    video_url: Optional[str] = None
    thumbnail_url: Optional[str] = None
    player_settings: Optional[PlayerSettings] = None

class VideoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    video_url: str
    thumbnail_url: Optional[str] = None
    duration: float
    plays_count: Optional[int] = 0
    player_settings: Dict[str, Any]
    created_at: datetime
    updated_at: datetime

class AnalyticsEventCreate(BaseModel):
    event_type: str = Field(..., description="impression, play, progress_25, progress_50, progress_75, progress_100, click")
    watch_time_seconds: Optional[float] = 0.0
    session_id: Optional[str] = None
    referer: Optional[str] = None

class HourlyMetric(BaseModel):
    hour: int
    label: str
    impressions: int
    plays: int
    clicks: int = 0

class PeakHour(BaseModel):
    hour: int
    label: str
    impressions: int
    plays: int
    total_activity: int

class VideoMetricsResponse(BaseModel):
    video_id: str
    period: Optional[str] = "all"
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    timezone: str = "America/Sao_Paulo (Horário de Brasília / UTC-3)"
    total_impressions: int
    unique_impressions: int
    total_plays: int
    unique_plays: int
    play_rate: float
    total_clicks: int
    ctr: float
    avg_watch_time_seconds: float
    retention: Dict[str, int]
    hourly_distribution: List[HourlyMetric] = []
    peak_hour: Optional[PeakHour] = None


class BulkDeleteRequest(BaseModel):
    video_ids: List[str] = Field(..., min_length=1)


class BulkDeleteResponse(BaseModel):
    deleted_count: int
    deleted_ids: List[str]

File: frontend/src/types/video.ts
export type SmartAutoplaySize = 'mini' | 'small' | 'medium' | 'large'

export interface SmartAutoplaySettings {
  enabled: boolean
  mode?: 'smart' | 'direct'
  text?: string
  subtext?: string
  button_color?: string
  button_text?: string
  restart_on_unmute?: boolean
  size?: SmartAutoplaySize
}

export interface FloatingPlayerSettings {
  enabled: boolean
  position?: 'bottom-right' | 'bottom-left'
  width?: number
  closeable?: boolean
}

export interface PitchDelaySettings {
  enabled: boolean
  time: number
  target_css_selector?: string
  auto_scroll?: boolean
  scroll_offset?: number
  persistence?: boolean
}

export interface PixelEventConfig {
  trigger: 'percent_25' | 'percent_50' | 'percent_75' | 'percent_100' | 'pitch'
  event_name: string
  enabled: boolean
}

export interface TrackingPixelsSettings {
  enabled: boolean
  facebook_pixel_id?: string
  google_analytics_id?: string
  tiktok_pixel_id?: string
  events?: PixelEventConfig[]
}

export interface DomainProtectionSettings {
  enabled: boolean
  allowed_domains?: string[]
  anti_download?: boolean
}

export interface PlayerControlsConfig {
  rewind_10s?: boolean
  forward_10s?: boolean
  volume?: boolean
  fullscreen?: boolean
  speed_control?: boolean
  progress_bar?: boolean
  video_time?: boolean
}

export interface ChapterItem {
  id: string
  time: string
  seconds: number
  title: string
}

export interface ChaptersSettings {
  enabled: boolean
  items: ChapterItem[]
}

export interface PlayerSettings {
  primary_color: string
  autoplay: boolean
  show_controls: boolean
  border_radius?: number
  aspect_ratio?: '16:9' | '9:16'
  cta_enabled: boolean
  cta_time: number
  cta_text: string
  cta_link: string
  play_button_shape?: 'circle' | 'rounded' | 'square' | 'minimal'
  play_button_size?: 'small' | 'medium' | 'large'
  default_width?: string
  default_ratio?: string
  turbo_enabled?: boolean
  playback_rate?: number
  smart_autoplay?: SmartAutoplaySettings
  floating_player?: FloatingPlayerSettings
  pitch_delay?: PitchDelaySettings
  tracking_pixels?: TrackingPixelsSettings
  domain_protection?: DomainProtectionSettings
  controls_config?: PlayerControlsConfig
  chapters?: ChaptersSettings
  transparent_background?: boolean
  remove_black_bars?: boolean
  fit_mode?: 'cover' | 'contain'
}


export interface Video {
  id: string
  title: string
  video_url: string
  thumbnail_url?: string
  duration: number
  plays_count?: number
  player_settings: PlayerSettings
  created_at: string
  updated_at: string
}

export interface HourlyMetric {
  hour: number
  label: string
  impressions: number
  plays: number
  clicks?: number
}

export interface PeakHour {
  hour: number
  label: string
  impressions: number
  plays: number
  total_activity: number
}

export interface VideoMetrics {
  video_id: string
  period?: string
  start_date?: string | null
  end_date?: string | null
  total_impressions: number
  unique_impressions?: number
  total_plays: number
  unique_plays?: number
  play_rate: number
  total_clicks: number
  ctr: number
  avg_watch_time_seconds: number
  retention: {
    '25%': number
    '50%': number
    '75%': number
    '100%': number
  }
  hourly_distribution?: HourlyMetric[]
  peak_hour?: PeakHour | null
}
